### PR TITLE
feat: llama.cpp gRPC C++ backend

### LINKED ISSUE #944: LLAMA_MAX_DEVICES bug
error report :stderr GGML_ASSERT: binding.cpp:712: split_arg.size() <= LLAMA_MAX_DEVICES

I find that default LLAMA_MAX_DEVICES = 1.LocalAI can not accelerate llama2 in multiple gpus when use tensor_split param. Please fix this bug.

### LINKED ISSUE #1154: [backend] llama-cpp, C++ gRPC backend
**Is your feature request related to a problem? Please describe.**
keeping in sync the golang binding can lag behind for certain reasons related to the golang toolchain. It'd be easier to also have a c++ implementation that requires less  maitenance and so can be bumped almost automatically

**Describe the solution you'd like**
At this point we can also have a pure C++ llama-cpp gRPC server binding with just the few little things we need to add on top

**Describe alternatives you've considered**

**Additional context**
This would allow us to open up bugs upstream more easily as we have a less heavy implementation on top. We could also upstream the gRPC server

### PR DESCRIPTION
**Pull request by mudler**
**Description**

This PR fixes #1154,  fixes https://github.com/go-skynet/LocalAI/issues/1017 and fixes #944 

It provides a C++ gRPC server to use in place of the go-llama binding. note it is much more flexible as now the Makefile controls directly the commit/branch/tag of llama.cpp. There are of course lots of optimizations that can be done, but this provides the first steps.

There are still some TODOs some required for a feature parity with the golang backend:

- Expose lora parameters
- Embeddings (feat parity)
- Speculative sampling (feat parity)
- Parallel sampling
- Additionally expose things like logits bias and an infill endpoint

What supposedly should work:

Everything, except loras. Including also CUDA/etc support

Next:

- Merge it
- Merge and unblock #1116 so we test with GPUs automatically too
- expose lora (now it needs also a scale factor)
- move golang backends inside `backends` (nearby the `cpp` one)
- move `extra` inside `backends`
- make it feature-parity with the current `llama backend`
- sunset the current `llama` in favor of this one 
- rename `llama-stable` to `llama-ggml` to indicate support for old ggml file format (prior to gguf)

### CODE REVIEW

**Comment by mudler** [comment]
reminder to myself: create issues for follow-ups that are not tackled here
